File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Pair<Integer,Integer>>stack = new Stack<>();
        int max_area =0;
        int n = heights.length;
        for(int i=0;i<n;i++){
            int min_index = i;
            while(!stack.isEmpty() && stack.peek().getKey() > heights[i]){
                max_area = Math.max(max_area, stack.peek().getKey() * (i-stack.peek().getValue()));
                min_index = Math.min(min_index, stack.peek().getValue());
                stack.pop();

            }
            stack.add(new Pair<>(heights[i], min_index));
        }
        while(!stack.isEmpty()){
            max_area = Math.max(max_area, stack.peek().getKey() * (n-stack.peek().getValue()));
            stack.pop();
        }
        return max_area;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int,int>>stack;
        int max_area=0;
        int n = heights.size();
        for(int i=0;i<heights.size();i++){
            int ind=-1;
            while(!stack.empty() && stack.top().second > heights[i]){
                int idx= stack.top().first;
                int ht=stack.top().second;
                stack.pop();
                int area = ht*(i-idx);
                max_area = max(max_area, area);
                ind=idx;

            }
            if(ind==-1){
                stack.push({i,heights[i]});
            }else{
                stack.push({ind,heights[i]});
            }
        }
        while(!stack.empty()){
            int idx= stack.top().first;
            int ht=stack.top().second;
            stack.pop();
            int area = ht*(n-idx);
            max_area = max(max_area, area);
        }
        return max_area;
    }
};